core/src/main/java/com/linecorp/armeria/common/stream/SubscribeOnStreamMessage.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class SubscribeOnStreamMessage<T> implements StreamMessage<T> {
+
+    private final StreamMessage<T> upstream;
+    private final EventExecutor upstreamExecutor;
+
+    SubscribeOnStreamMessage(StreamMessage<T> upstream, EventExecutor upstreamExecutor) {
+        this.upstream = upstream;
+        this.upstreamExecutor = upstreamExecutor;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return upstream.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return upstream.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return upstream.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return upstream.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor downstreamExecutor,
+                          SubscriptionOption... options) {
+        if (upstreamExecutor.inEventLoop()) {
+            if (upstreamExecutor == downstreamExecutor) {
+                upstream.subscribe(subscriber, downstreamExecutor, options);
+            } else {
+                upstream.subscribe(new SchedulingSubscriber<>(downstreamExecutor, subscriber),
+                                   upstreamExecutor, options);
+            }
+        } else {
+            upstreamExecutor.execute(() -> upstream.subscribe(
+                    new SchedulingSubscriber<>(downstreamExecutor, subscriber),
+                    upstreamExecutor, options));
```

Comment:
I think we can probably do:
```suggestion
        final Subscriber<? super T> subscriber0;
        if (upstreamExecutor == downstreamExecutor) {
            subscriber0 = subscriber;
        } else {
            subscriber0 = new SchedulingSubscriber<>(downstreamExecutor, subscriber);
        }
        if (upstreamExecutor.inEventLoop()) {
            upstream.subscribe(subscriber0, downstreamExecutor, options);
        } else {
            upstreamExecutor.execute(() -> upstream.subscribe(subscriber0, upstreamExecutor, options));
        }
```